src/pfe/portal/modules/Templates.js

```diff
@@ -115,26 +138,97 @@ module.exports = class Templates {
     return this.projectTemplates;
   }
 
-  async getTemplatesOfStyle(projectStyle) {
-    const allTemplates = await this.getTemplateList();
-    const relevantTemplates = allTemplates.filter(template =>
-      getTemplateStyle(template) === projectStyle
+  // Save the default list to disk so the user can potentially edit it (WHEN CODEWIND IS NOT RUNNING)
+  async writeRepositoryList() {
+    await fs.writeJson(this.repositoryFile, this.repositoryList, { spaces: '  ' });
+    log.info(`Repository list updated.`);
+  }
+
+  getRepositories() {
+    return this.repositoryList;
+  }
+
+  getEnabledRepositories() {
+    return this.getRepositories().filter(repo =>
+      // if the repo doesn't specify whether it's enabled, consider it enabled
+      (repo.enabled || !repo.hasOwnProperty('enabled'))
     );
-    return relevantTemplates;
   }
 
-  writeRepositoryList() {
-    // Use a callback here so we don't block the response to the request.
-    fs.writeJson(this.repositoryFile, this.repositoryList, { spaces: '  ' }, err => {
-      if (err) {
-        log.error(`Error writing repository list to ${this.repositoryFile}: ${err}`)
-      }
-      log.info(`Repository list updated.`);
-    });
+  doesRepositoryExist(repoUrl) {
+    try {
+      this.getRepository(repoUrl);
+      return true;
+    } catch (error) {
+      return false;
+    }
   }
 
-  getRepositories() {
-    return this.repositoryList;
+  getRepositoryIndex(url) {
+    const repos = this.getRepositories();
+    const index = repos.findIndex(repo => repo.url === url);
+    return index;
+  }
+
+  /**
+   * @param {String} url
+   * @return {JSON} reference to the repo object in this.repositoryList
+   */
+  getRepository(url) {
+    const index = this.getRepositoryIndex(url);
+    if (index < 0) throw new Error(`no repository found with URL '${url}'`);
+    const repo = this.getRepositories()[index];
+    return repo;
+  }
+
+  async enableRepository(url) {
+    const repo = this.getRepository(url);
+    repo.enabled = true;
+    await this.writeRepositoryList();
+  }
+
+  async disableRepository(url) {
+    const repo = this.getRepository(url);
+    repo.enabled = false;
+    await this.writeRepositoryList();
+  }
+
+  async performOperation(operation) {
+    const { op, url, value } = operation;
+    let operationResult = {};
+    if (op === 'enable') {
+      operationResult = await this.performEnableOrDisableOperation({ url, value });
+    }
+    operationResult.requestedOperation = operation;
+    return operationResult;
+  }
+
+  /**
+   * @param {JSON} { url (URL of template repo to enable or disable), value (true|false)}
+   * @returns {JSON} { status, error (optional) }
+   */
+  async performEnableOrDisableOperation({ url, value }) {
+    if (!this.doesRepositoryExist(url)) {
+      return {
+        status: 404,
+        error: 'Unknown repository URL',
+      };
+    }
+    try {
+      if (value === 'true') {
+        await this.enableRepository(url);
```

Comment:
I might be off because it's kind of hard to understand the code without pulling it down, but it looks to me like if we do a batch enable/disable, the repo list file gets written out to once for every updated repo. IE, if we update 5 repos, the file gets written to 5 times. 

Putting aside that this is not optimized, is it possible that we'll get FS errors trying to read/write the file that is already open? Or is there some locking mechanism in place? I see that you are using Promise.all to do the batch update, so it looks like they will all execute in parallel.

Might it be better to write to the file just once per API call? Anyway if I'm way off, let me know.

Reply:
I've changed the code to write to the file just once per API call. Can look at locking the file next iteration